Firestore/src/FirestoreSessionHandler.php

```diff
@@ -0,0 +1,265 @@
+ *
+ * $handler = new FirestoreSessionHandler($firestore);
+ *
+ * session_set_save_handler($handler, true);
+ * session_save_path('sessions');
+ * session_start();
+ *
+ * // Then read and write the $_SESSION array.
+ *
+ * ```
+ *
+ * The above example automatically writes the session data. It's handy, but
+ * the code doesn't stop even if it fails to write the session data, because
+ * the `write` happens when the code exits. If you want to know the session
+ * data is correctly written to the Firestore, you need to call
+ * `session_write_close()` explicitly and then handle `E_USER_WARNING`
+ * properly like the following example.
+ *
+ * Example with error handling:
+ *
+ * ```
+ * use Google\Cloud\Firestore\FirestoreClient;
+ *
+ * $firestore = new FirestoreClient(['projectId' => $projectId]);
+ *
+ * $handler = new FirestoreSessionHandler($firestore);
+ * session_set_save_handler($handler, true);
+ * session_save_path('sessions');
+ * session_start();
+ *
+ * // Then read and write the $_SESSION array.
+ *
+ * function handle_session_error($errNo, $errStr, $errFile, $errLine) {
+ *     # We throw an exception here, but you can do whatever you need.
+ *     throw new Exception("$errStr in $errFile on line $errLine", $errNo);
+ * }
+ * set_error_handler('handle_session_error', E_USER_WARNING);
+ * // If `write` fails for any reason, an exception will be thrown.
+ * session_write_close();
+ * restore_error_handler();
+ * // You can still read the $_SESSION array after closing the session.
+ * ```
+ * @see http://php.net/manual/en/class.sessionhandlerinterface.php SessionHandlerInterface
+ */
+class FirestoreSessionHandler implements SessionHandlerInterface
+{
+    /* @var int */
+    private $gcLimit;
+    /* @var FirestoreClient */
+    private $firestore;
+    /* @var CollectionReference */
+    private $collection;
+
+    /**
+     * Create a custom session handler backed by Cloud Firestore.
+     *
+     * @param FirestoreClient $firestore Firestore client.
+     * @param int $gcLimit [optional] A number of entities to delete in the
+     *        garbage collection.  Defaults to 0 which means it does nothing.
+     *        The value larger than 1000 will be cut down to 1000.
+     */
+    public function __construct(
+        FirestoreClient $firestore,
+        $gcLimit = 0
+    ) {
+        $this->firestore = $firestore;
+        // Cut down to 1000
+        $this->gcLimit = min($gcLimit, 1000);
+    }
+
+    /**
+     * Start a session, by getting the Firebase collection from $savePath.
+     *
+     * @param string $savePath The value of `session.save_path` setting will be
+     *        used here as the Firestore collection ID.
+     * @param string $sessionName The value of `session.name` setting will be
+     *        used here as the Firestore document ID prefix (ex: "PHPSESSID:").
+     * @return bool
+     */
+    public function open($savePath, $sessionName)
+    {
+        if (false !== strpos($savePath, '/')) {
```

Comment:
The intention is to just make sure '/' doesn't exist in the collection name, as this is an invalid character for collections.

I agree that it may be fine to just let the client throw an exception.

Reply:
Unfortunately since we mock these methods, it's not something we can really test. Is that okay?

Reply:
I think so! We have test coverage of the client, so omitting the test here should be fine.

Reply:
Also, I am not sure how to test the `runTransaction` closure with the mock library implementation. Do you have an example of testing something like this anywhere?

Reply:
OK gave it a crack. PTAL.